src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+      restriction_opt.substr(pointer_name_end + 1);
+    auto const target_name_strings = split_string(target_names_substring, ',');
+    if(target_name_strings.size() == 1 && target_name_strings[0].empty())
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "missing target list for function pointer restriction " + pointer_name,
+        "--" RESTRICT_FUNCTION_POINTER_OPT,
+        restriction_format_message};
+    }
+    auto const target_names = ([&target_name_strings] {
+      auto target_names = std::unordered_set<irep_idt>{};
+      std::transform(
+        target_name_strings.begin(),
+        target_name_strings.end(),
+        std::inserter(target_names, target_names.end()),
+        string2id);
+      return target_names;
+    })();
+    for(auto const &target_name : target_names)
+    {
+      if(target_name == ID_empty_string)
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "leading or trailing comma in restrictions for `" + pointer_name +
+            "'",
+          "--" RESTRICT_FUNCTION_POINTER_OPT,
+          restriction_format_message);
+      }
+    }
+    if(!function_pointer_restrictions
+          .emplace(irep_idt{pointer_name}, target_names)
+          .second)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "function pointer restriction for `" + pointer_name +
+          "' was specified twice",
+        "--" RESTRICT_FUNCTION_POINTER_OPT};
+    };
+  }
+  return function_pointer_restrictions;
+}
+
+function_pointer_restrictionst::restrictionst
+parse_function_pointer_restrictions_from_file(
+  const std::list<std::string> &filenames,
+  message_handlert &message_handler)
+{
+  auto merged_restrictions = function_pointer_restrictionst::restrictionst{};
+  for(auto const &filename : filenames)
+  {
+    auto const restrictions =
+      function_pointer_restrictionst::read_from_file(filename, message_handler);
+    merged_restrictions = merge_function_pointer_restrictions(
+      merged_restrictions, restrictions.restrictions);
+  }
+  return merged_restrictions;
+}
+} // namespace
+
+function_pointer_restrictionst function_pointer_restrictionst::from_options(
+  const optionst &options,
+  message_handlert &message_handler)
+{
+  auto const restriction_opts =
+    options.get_list_option(RESTRICT_FUNCTION_POINTER_OPT);
+  auto const commandline_restrictions =
+    parse_function_pointer_restrictions_from_command_line(restriction_opts);
+  auto const file_restrictions = parse_function_pointer_restrictions_from_file(
+    options.get_list_option(RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT),
+    message_handler);
+  return {merge_function_pointer_restrictions(
+    file_restrictions, commandline_restrictions)};
+}
+
+function_pointer_restrictionst function_pointer_restrictionst::read_from_file(
+  const std::string &filename,
+  message_handlert &message_handler)
+{
+  auto failed = [](bool failFlag) { return failFlag; };
```

Comment:
~:question:~ :pick:  This name seems a little misleading - is this not just the identity function?

Reply:
Ah I see what is being done with this... A comment might not be an awful idea, but not blocking

Reply:
Removed this lambda.